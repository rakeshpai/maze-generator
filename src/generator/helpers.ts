import { Lens } from 'monocle-ts';
import { Cell, Maze, locations, Location, CellWithLocation } from './types';

// Local helpers
const range = (max: number) =>
  (new Array(max)).fill(0).map((unused, index) => index);

const populateList = <T>(
  width: number, height: number, populator: (x: number, y: number) => T
): T[] => range(width).reduce<T[]>(
  (acc, x) => acc.concat(range(height).map(y => populator(x, y))),
  []
);

const inc = (x: number) => x + 1;
const dec = (x: number) => x - 1;
const id = <T>(x: T) => x;

export const pickRandomlyFromList = <T>(list: T[]) =>
  list[Math.floor(Math.random() * list.length)];

export const oppositeLocation = (location: Location) =>
  locations[(locations.indexOf(location) + 2) % locations.length];

const neighboursDefinition: [Location, (x: number) => number, (y: number) => number][] = [
  ['top', dec, id],
  ['left', id, dec],
  ['right', id, inc],
  ['bottom', inc, id]
];

const matchingCell = (x: number, y: number) => (cell: Cell) =>
  cell.x === x && cell.y === y;

// Lenses
const wallLens = (location: Location) => Lens.fromPath<Cell>()(['walls', location]);

const wallLenses: { [key in Location]: Lens<Cell, boolean> } = {
  top: wallLens('top'),
  right: wallLens('right'),
  bottom: wallLens('bottom'),
  left: wallLens('left')
}

const cellLens = (x: number, y: number) => new Lens<Maze, Cell>(
  maze => maze.cells.find(matchingCell(x, y))!,
  cell => maze => ({
    ...maze,
    cells: maze.cells.map(c => matchingCell(x, y)(c) ? cell : c)
  })
);

const currentCellLens = Lens.fromProp<Maze>()('currentCell');
const visitedLens = Lens.fromProp<Cell>()('visited');

// Maze helpers
export const initializeMaze = (width: number, height: number): Maze => {
  const cells = populateList(width, height, (x, y): Cell => ({
    x, y,
    walls: { top: true, right: true, bottom: true, left: true},
    visited: false
  }));

  return {
    width, height, cells,
    stack: [], currentCell: [ cells[0].x, cells[0].y ]
  };
};

export const hasUnvisitedCells = (maze: Maze) =>
  maze.cells.some(cell => !cell.visited);

export const pushCurrentCellToStack = (maze: Maze) =>
  ({ ...maze, stack: [ maze.currentCell, ...maze.stack ]})
  
export const popFromStackToCurrentCell = (maze: Maze) =>
  ({ ...maze, stack: maze.stack.slice(1), currentCell: maze.stack[0] });

const cellAt = (maze: Maze, x: number, y: number) => {
  if(x < 0 || x > maze.height - 1 || y < 0 || y > maze.width - 1) {
    return undefined;
  }

  return maze.cells[ (maze.width * x) + y ];
};

export const setCurrentCell = (x: number, y: number) =>
  currentCellLens.set([x, y]);

export const markCellAsVisited = (x: number, y: number) =>
  cellLens(x, y).compose(visitedLens).set(true);

const neighboursOfCurrentCell = (maze: Maze) =>
  neighboursDefinition.reduce<CellWithLocation[]>(
    (acc, [location, xOp, yOp]) => {
      const cell = cellAt(maze, xOp(maze.currentCell[0]), yOp(maze.currentCell[1]));
      if(!cell) return acc;

      return acc.concat({ ...cell, location })
    },
    []
  );

export const unvisitedNeighboursForCurrentCell = (maze: Maze) =>
  neighboursOfCurrentCell(maze)
    .filter((cell: CellWithLocation) => !cell.visited);

const removeCellWall = (cell: Cell, location: Location) =>
  cellLens(cell.x, cell.y).compose(wallLenses[location]).set(false);

export const removeWallBetweenCurrentCellAnd = (cell: CellWithLocation) => (maze: Maze): Maze => {
  const currentCell = cellLens(...maze.currentCell).get(maze);

  const tempMaze = removeCellWall(currentCell, cell.location)(maze);
  return removeCellWall(cell, oppositeLocation(cell.location))(tempMaze);
};

export const isCurrent = (cell: Cell) => (maze: Maze) => {
  const [x, y] = currentCellLens.get(maze)
  return cell.x === x && cell.y === y;
};
