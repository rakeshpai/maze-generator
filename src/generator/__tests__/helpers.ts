import { initializeMaze, oppositeLocation, unvisitedNeighboursForCurrentCell, setCurrentCell, markCellAsVisited, removeWallBetweenCurrentCellAnd } from "../helpers";

it('should create a starting maze', () => {
  const maze = initializeMaze(10, 10);
  
  expect(maze.cells.length).toBe(100);
  expect(maze.width).toBe(10);
  expect(maze.height).toBe(10);
  expect(maze.currentCell[0]).toBe(0);
  expect(maze.currentCell[1]).toBe(0);
  expect(maze.stack).toEqual([]);

  for(let i = 0; i<10; i++) {
    for(let j=0; j<10; j++) {
      const cell = maze.cells[ (i * 10) + j ];
      expect(cell.x).toBe(i);
      expect(cell.y).toBe(j);
    }
  }
});

it('should return the opposite location', () => {
  expect(oppositeLocation('top')).toBe('bottom');
  expect(oppositeLocation('right')).toBe('left');
  expect(oppositeLocation('bottom')).toBe('top');
  expect(oppositeLocation('left')).toBe('right');
});

it('should return unvisited neighbours', () => {
  const maze1 = initializeMaze(10, 10);
  expect(unvisitedNeighboursForCurrentCell(maze1)).toMatchSnapshot();

  const maze2 = setCurrentCell(5, 5)(maze1);
  expect(unvisitedNeighboursForCurrentCell(maze2)).toMatchSnapshot();

  const maze3 = setCurrentCell(4, 5)(markCellAsVisited(5, 5)(maze2));
  expect(unvisitedNeighboursForCurrentCell(maze3)).toMatchSnapshot();
});

it('should remove cell walls', () => {
  const maze1 = initializeMaze(10, 10);
  const rightCell = unvisitedNeighboursForCurrentCell(maze1)
    .find(c => c.location === 'right');
  
  const maze2 = removeWallBetweenCurrentCellAnd(rightCell!)(maze1);

  expect(maze2.cells[0].walls.right).toBe(false);
  expect(maze2.cells[1].walls.left).toBe(false);
});