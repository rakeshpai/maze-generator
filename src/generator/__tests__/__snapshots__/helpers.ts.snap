// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return unvisited neighbours 1`] = `
Array [
  Object {
    "location": "right",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 0,
    "y": 1,
  },
  Object {
    "location": "bottom",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 1,
    "y": 0,
  },
]
`;

exports[`should return unvisited neighbours 2`] = `
Array [
  Object {
    "location": "top",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 4,
    "y": 5,
  },
  Object {
    "location": "left",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 5,
    "y": 4,
  },
  Object {
    "location": "right",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 5,
    "y": 6,
  },
  Object {
    "location": "bottom",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 6,
    "y": 5,
  },
]
`;

exports[`should return unvisited neighbours 3`] = `
Array [
  Object {
    "location": "top",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 3,
    "y": 5,
  },
  Object {
    "location": "left",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 4,
    "y": 4,
  },
  Object {
    "location": "right",
    "visited": false,
    "walls": Object {
      "bottom": true,
      "left": true,
      "right": true,
      "top": true,
    },
    "x": 4,
    "y": 6,
  },
]
`;
